### PR TITLE
fix(core): Enhanced Fetch improvements

### PR DESCRIPTION
**Pull request by eirikurn**
## What

* Check cache before circuit breaker and autoAuth (token exchange or client credentials).
* Add cache status to error logging.
* Add cache tag to metrics.

* Configure our NestJS servers to allow 30 seconds of idle timeout.
* Configure our clients to reuse connections with default 10 second idle timeout.
* Includes E2E tests for Enhanced Fetch.

## Why

* To optimise EnhancedFetch when both cache and auto auth are configured. 
* So we can provide partial service from cache even though the circuit breaker has been tripped.
* To have cache metrics in datadog.

* To provide even better performance during load by reusing connections and skipping DNS, TCP and even TLS setup roundtrips.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

**Review by pshomov — APPROVED** [review]
LGTM

File: libs/clients/middlewares/src/lib/e2e/enhancedFetch.spec.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        },
+      })
+    })
+    server.on('error', (err) => {
+      reject(err)
+    })
+    server.listen()
+  })
+
+const sleep = (ms: number): Promise<void> =>
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/clients/middlewares/src/lib/e2e/enhancedFetch.spec.ts:62]: There are some issues with fake timers and promises which I was unable to get around. :(

Refactored sleep into `@island.is/shared/utils`.

File: libs/clients/middlewares/src/lib/withCache/withCache.ts [hunk 2]
```diff
@@ -1,8 +1,10 @@
 import CachePolicy from 'http-cache-semantics'
 
 import { FetchAPI, Headers, HeadersInit, Request, Response } from '../nodeFetch'
+import { FetchError } from '../FetchError'
 import { CacheEntry, CacheMiddlewareConfig, CachePolicyInternal } from './types'
 import { CacheResponse } from './CacheResponse'
+import { serializeCacheStatusHeader } from './CacheStatus'
```
Comment (saevarma) [inline libs/clients/middlewares/src/lib/withCache/withCache.ts:7]: GitHub found unused import

**Review by saevarma — APPROVED** [review]
LGTM.
Removed `automerge` if you want to remove the unused import before merging.

**Pull request merged**